File: docs/snippets/all/archetypes/encoded_depth_image.py
"""Log an encoded depth image stored as a 16-bit PNG or RVL file."""

import sys
from pathlib import Path

import dalaran as dl

if len(sys.argv) < 2:
    print(
        f"Usage: {sys.argv[0]} <path_to_depth_image.[png|rvl]>", file=sys.stderr
    )
    sys.exit(1)

depth_path = Path(sys.argv[1])

dl.init("dalaran_example_encoded_depth_image", spawn=True)

depth_png = depth_path.read_bytes()
if depth_path.suffix.lower() == ".png":
    media_type = dl.components.MediaType.PNG
else:
    media_type = dl.components.MediaType.RVL

dl.log(
    "depth/encoded",
    dl.EncodedDepthImage(
        blob=depth_png,
        media_type=media_type,
        meter=0.001,
    ),
)

File: docs/snippets/all/archetypes/encoded_depth_image.cpp
//! Log an encoded depth image stored as a 16-bit PNG or RVL file

#include <dalaran.hpp>

#include <filesystem>
#include <fstream>
#include <iostream>
#include <vector>

namespace fs = std::filesystem;

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <path_to_depth_image.[png|rvl]>"
                  << std::endl;
        return 1;
    }

    const auto rec =
        dalaran::RecordingStream("dalaran_example_encoded_depth_image");
    rec.spawn().exit_on_failure();

    const auto depth_path = fs::path(argv[1]);
    std::ifstream file(depth_path, std::ios::binary);
    if (!file) {
        std::cerr << "Failed to open encoded depth image: " << depth_path
                  << std::endl;
        return 1;
    }

    std::vector<uint8_t> bytes{
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>()
    };
    // Determine media type based on file extension
    dalaran::MediaType media_type;
    if (depth_path.extension() == ".png") {
        media_type = dalaran::MediaType::png();
    } else {
        media_type = dalaran::MediaType::rvl();
    }

    rec.log(
        "depth/encoded",
        dalaran::archetypes::EncodedDepthImage()
            .with_blob(dalaran::components::Blob(
                dalaran::Collection<uint8_t>::take_ownership(std::move(bytes))
            ))
            .with_media_type(media_type)
            .with_meter(0.001f)
    );
}
